### PR TITLE
Initial WMArchive REST service

### PR DESCRIPTION
**Pull request by vkuznet**
Please take initial implementation of WMArchive REST service (proxy server). Here is its configuration file:

```
import socket
from WMCore.Configuration import Configuration
from os import path

HOST = socket.gethostname().lower()
ROOTDIR = __file__.rsplit('/', 3)[0]
config = Configuration()

main = config.section_("main")
srv = main.section_("server")
srv.thread_pool = 30
main.application = "wmarchive"
main.port = 8246 # main application port it listens on
#main.index = "ui"
main.index = "data"
# Defaults to allow any CMS authenticated user. Write APIs should require
# additional roles in SiteDB (i.e. "Admin" role for the "ReqMgr" group)
#main.authz_defaults = {"role": None, "group": None, "site": None}

#sec = main.section_("tools").section_("cms_auth")
#sec.key_file = "%s/auth/wmcore-auth/header-auth-key" % ROOTDIR
#sec.key_file = "/Users/vk/CMS/DMWM/GIT/WMCore/auth-key"

# this is where the application will be mounted, where the REST API
# is reachable and this features in CMS web frontend rewrite rules
app = config.section_(main.application) # string containing "reqmgr2"
app.admin = "cms-service-webtools@cern.ch"
app.description = "CMS data operations WMArchive."
app.title = "CMS Request Manager (WMArchive)"

views = config.section_("views")

# practical to have this kind of configuration values not in
# service related RPM (difficult/impossible to change in CMS web
# deployment) but in the deployment configuration for the service

# redirector for the REST API implemented handlers
data = views.section_("data")
data.object = "WMCore.WMArchive.Service.RestApiHub.RestApiHub"
```

and (assuming all environment is set) it can be run as following:

```
wmc-httpd -r -d /tmp -l wma.log ./wmarch_config.py
```

### CODE REVIEW

File: src/python/WMCore/WMArchive/Service/RestApiHub.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        :arg str mount: API URL mount point; passed to all entities."""
+        
+        RESTApi.__init__(self, app, config, mount)
+        
+        cherrypy.log("WMArchive entire configuration:\n%s" % Configuration.getInstance())    
+        cherrypy.log("WMArchive REST hub configuration subset:\n%s" % config)
+        
+        self._add({
+            "info": WMAInfo(app, self, config, mount),
+            "write": WMAPost(app, self, config, mount),
```
Comment (ticoann) [inline src/python/WMCore/WMArchive/Service/RestApiHub.py]: I think the name should be something related to the data not about the action ("write")

File: src/python/WMCore/WMArchive/Service/RestApiHub.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            "fetch": WMAGet(app, self, config, mount),
```
Comment (ticoann) [inline src/python/WMCore/WMArchive/Service/RestApiHub.py]: Here as well

File: src/python/WMCore/WMArchive/Web/WMArchiveService.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        cherryconf.update({'tools.encode.on': True,
+                           'tools.gzip.on': True,
+                           'tools.gzip.mime_types': mime_types,
+                          })
+
+        # initialize rest API
+        statedir = '/tmp'
+        app = RESTMain(config, statedir) # REST application
+        mount = '/rest' # mount point for cherrypy service
+        api = RestApiHub(app, config.mgr, mount)
```
Comment (ticoann) [inline src/python/WMCore/WMArchive/Web/WMArchiveService.py]: The way mounting new REST class is a different from WMCore/WebTools,  Although I don't think you need to follow exactly but you can check how it is done in sitedb
i.e.)
Configuration is pointing which object to construct the front page.
https://github.com/dmwm/sitedb/blob/master/src/python/Config.py#L32

Front page construction
https://github.com/dmwm/sitedb/blob/master/src/python/FrontPage.py

File: src/python/WMCore/WMArchive/Service/Methods.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        validate_str("query", param, safe, PAT_EXAMPLE, optional=True)
+
+    @restcall(formats = [('application/json', JSONFormat())])
+    @tools.expires(secs=-1)
+    def get(self, query):
+        "GET request with given query, all the work is done by WMArchiveManager"
+        docs = self.mgr.read(query)
+        return docs
+
+class WMAPost(RESTEntity):
```
Comment (ticoann) [inline src/python/WMCore/WMArchive/Service/Methods.py]: Maybe we can have more data oriented name, I think the verb can define the operation.

File: src/python/WMCore/WMArchive/Service/Methods.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        return True
+
+    @restcall
+    @tools.expires(secs=-1)
+    def get(self):
+        wmcore_reqmgr_version = WMCore.__version__
+        result = {"WMArchive version": self.mgr.version()}
+        return rows([result])
+
+class WMAGet(RESTEntity):
```
Comment (ticoann) [inline src/python/WMCore/WMArchive/Service/Methods.py]: Maybe we can have more data oriented name, I think the verb can define the operation.

**Comment by ticoann** [comment]
Hi Valentin, as Yuyi mentioned I think it might be better the code is hosted in different repository.
https://github.com/dmwm/WMArchive
Since this project will contain different set of dependencies than other REST applications (ReqMgr2, WMStats) and I expect this will be much bigger and complicated application than those.

In terms of the code, in general,
RESTEntity is designed to operates over specific data. It will be implemented one or more verbs on that that. (i.e. POST, PUT, GET, DELETE)
I think in some cases it doesn't really fits the use case. (I had some problems applying that concept on reqmgr2).
Anyway, this is Lassi's design concept, so sitedb will be the good example how he initially envision this.

https://github.com/dmwm/sitedb/blob/master/src/python/DataPeople.py

I added some in line comment

**Comment by vkuznet** [comment]
I migrated code (and fixed it with respect of provided comments) to WMArchive repository. The new PR is https://github.com/dmwm/WMArchive/pull/1 I'll keep this open until we merge new PR. So far I only modified data classes, but didn't yet change service wrt sitedb layout. I need to think about it if it is appropriate here. This can be done later. Please review new PR.

**Comment by ticoann** [comment]
This is moved to WMArchive repository, closing it

**Pull request closed without merging**